# Parameters for TacTip Contact detection on TacTip
# ===========================================================
# - Tactip camera description
# ===========================================================
tactip:
    camera:                             # TacTip camera intrinsic parameter        
        depth: 48                       # distance camera lens to tip [mm]
        fx: 5.5232396906481722e+02      # - x focal lenght
        fy: 5.5575248970574751e+02      # - y focal lenght
        cx: 3.1042034270984652e+02      # - x camera center
        cy: 2.4710324279132291e+02      # - y camera center
            
    markers: 127                         # Number of markes

    mm2pxl: 10                          # conversion millimeters to pixel in a frame

# ===========================================================
# - Image processing parameters
# ===========================================================
image_processing:     
    shape:                          # Reshaped image dimension
        width: 640
        height: 480

    mask: circle                   # Image mask  
    
    blur_kernel_size: 5             # Image noise filtering (Gaussian) kernel size  
    
    circle_recog:                   # Hough Gradient Circle recogition params
        min_radius: 1               #   - min circle radius
        max_radius: 10              #   - min circle radius
        param1: 100                 #   - Upper threshold for the internal Canny edge detecto
        param2: 5                   #   - Threshold for center detection
        min_distance: 10            #   - min distance of adjacent circle    


    blob_recog:                     # Determinant of Heussian Blob's recogition params (TacTip: 6.5,8.5,0.003, DigiTac: ??)
        min_sigma: 5                #   - minimum blob width
        max_sigma: 7.5              #   - maximum blob width
        threshold: 0.05             #   - intensity changes threshold
        


# ===========================================================
# - Gaussian Kernel Density Estimation algorithm parameters
# ===========================================================
markers_density:            # Params for Gaussian Kernel Algorithm
    rate: 50                # rate [Hz]  
    gaussian_kernel:        # Density Estimator Gaussian Kernel 
        h: 15               #   - Density Estimator Gaussian Kernel std. dev, (TacTip: 15, DigiTac: 30)  
                            #     (it should be close to mean distance in pixels between adjacent markers)                                    
        verbose: True       #   - Print kernel estimation elapsed time 
    resolution: 1           # Resolution of pixels for Density estimation
                            # i.e. d(u,v) = density[u//resolution,v//resolution]
    threshold: 0.02         # Maximum density variation allowed in a pixel before contact detection [1/mm^2]
    fit:                    # Stiffness params for deformation estimation. def = h(vol)
        type: quadratic     # type of curve (i.e linear, quadratic, power, logarithmic)   
        #a: -0.2112         #   - linear parameter (norm to r)
        #b: 4.1014          #   - quadratic parameter (norm to r)
        #a: 0.00044         #   - linear parameter (no norm)
        #b: -7.426e-9       #   - quadratic parameter (no norm)
        #a: 1.0725           #   - linear parameter (in mm)
        #b:  -0.0500         #   - quadratic parameter (in mm)
        a: 0.76873
        b: 0.0073294


# ========================================================
# - Voronoi cell algorithm parameters
# ========================================================
markers_volume:         # Params for Voronoi Algorithm 
    rate: 30                    # rate [Hz]  
    voronoi:                    # Volume voronoi cell     
        vertices: [[305,50],    #   - List of points to delimit area with markers  
                   [480,153],   #    (they must be in clockwise or counterclockwise order) 
                   [480,343],   #
                   [305,440],   #
                   [130,343],   #
                   [130,153]]   #                     
        verbose: True           #   - Print area evaluation elapsed time 
    resolution: 1               # Resolution of pixels for Density estimation 
                                # i.e. d(u,v) = density[u//resolution,v//resolution]
    threshold: 50               # Maximum volume variation allowed in a pixel before contact detection [1/mm^2]
    fit:                        # Stiffness params for deformation estimate. def = a*vol+b*vol^2
        type: quadratic
        a: 1.179                #   - linear parameter 
        b: -0.09304             #   - quadratic parameter